walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -12,12 +12,243 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+/**
+* Class to schedule meetings based on attendees free times.
+*/
+
 package com.google.sps;
 
+import java.util.Arrays;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.PriorityQueue;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> meetingTimesForRequiredAttendees = new ArrayList<TimeRange>();
+    Collection<TimeRange> meetingTimesForOptionalAttendees = new ArrayList<TimeRange>();
+    final long meetingDuration = request.getDuration();
+
+    // Return empty list if the meeting last more than a day
+    if (TimeRange.WHOLE_DAY.duration() < meetingDuration) {
+      return meetingTimesForRequiredAttendees;
+    }
+    
+    // All day available if there are no scheduled events
+    if (events.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    // List of required attendees
+    final Collection<String> meetingAttendees = request.getAttendees();
+
+    //Get the times where the attendees are not free
+    final PriorityQueue<TimeRange> requiredAttendeesOccupiedTimes = 
+      getAttendeesOccupiedTimes(events, meetingAttendees, true);
+    final PriorityQueue<TimeRange> optionalAttendeesTimeList = 
+      getAttendeesOccupiedTimes(events, meetingAttendees, false);
+
+    /** 
+    * Look for time ranges for optional attendees 
+    * if required attendees all available all day 
```

Comment:
```suggestion
    * if required attendees are available all day 
```
Nit: Is this a typo?

Reply:
Also, I think the /** */ style of comments is used for JSDoc. If you have a multiline comment, then I think it's okay to use //. So like below:

// Look for time ranges for optional attendees
// if required attendees are available all day

Reply:
Thanks for the recommendation Aki!